AndroidSDKCore/src/main/java/com/leanplum/internal/RequestOld.java

```diff
@@ -276,43 +243,47 @@ public void onApiResponse(ApiResponseCallback apiResponse) {
     return args;
   }
 
-  private void saveRequestForLater(Map<String, Object> args) {
-    try {
+  /**
+   * Saves requests into database.
+   * Saving will be executed on background thread serially.
+   * @param args json to save.
+   */
+  private void saveRequestForLater(final Map<String, Object> args) {
+    OperationQueue.sharedInstance().addOperation(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          Context context = Leanplum.getContext();
+          if (context == null) {
+            return;
+          }
 
-      Context context = Leanplum.getContext();
-      if (context == null) {
-        return;
-      }
+          SharedPreferences preferences = context.getSharedPreferences(LEANPLUM,
```

Comment:
SharedPreferences are mostly thread safe, i.e it won't crash, just the last who write wins, and since this is executed on serial queue we can assume that only one runnable will be executed at a time.

The synchronization was on RequestOld which was causing another issue visible in Unity only, due to need to maintain 30+ fps refresh rate, every lock was causing main thread to hiccup.

Reply:
sounds good. I wanted to confirm. There are other places where "SharedPreferences" are getting updated or read, not using the serial queue. Wont that interfere with this logic not bring syncronized?

Reply:
It won't because they read/write different keys.